### ISSUE TITLE
Putting an empty value into a storage slot messes up gas calculation in Istanbul

### ISSUE THREAD

**Issue by jochem-brouwer** [ethereumjs/ethereumjs-vm#729]
If you put an empty value (i.e. `Buffer.from('0000000000000000000000000000000000000000000000000000000000000000', 'hex')`) into any storage slot, and then write to this storage slot via a `SSTORE` operation, then `sstoreCleanGasEIP2200` (5000 gas) gets deducted from the remaining gas, while `sstoreInitGasEIP2200` (20000 gas) should be deducted. This is because internally if you `putContractStorage` using this value, it will encode the entire buffer and put it in the state trie, while it should put `new Buffer('')` (the empty `Buffer`) in the trie. Besides this issue, it also calculates the wrong root hash of the MPT, due to wrongly encoded 0-byte-leading buffers.

**Comment by jochem-brouwer** [comment]
It seems that `StateManager.putContractStorage` accepts values of any length. This appears to be a feature?

**Comment by jochem-brouwer** [comment]
I propose that `StateManager.putContractStorage` gets an extra boolean argument `putSafe` which is `false` by default. If `true`, it converts any fed data to rightly padded data, or throws: i.e. if you input a `Buffer` with more than 32 bytes, it throws. If it has leading zero's, delete these zero bytes.

**Comment by jochem-brouwer** [comment]
Can we throw an error in V5 if this happens? This messes up so many things (gas calculation and the MPT root). I do not think we should allow this and loudly error if someone puts a non-32-byte-length key with value into the storage. Also we do not save 0-equivalent values into the storage, and we should pad the Buffers if they are not 32 byte-length (or will we strictly enforce values are 32 byte-length?). Thoughts? @holgerd77

**Comment by alcuadrado** [comment]
I think making the state manager stricter would be the best option. I have been bitten by this before. Not because of the gas, but when making tests for Buidler EVM.

**Comment by jochem-brouwer** [comment]
Yeah the problem with this is that it is not obvious why it yields unexpected results. If you throw at the `StateManager` if you try to use a Buffer with wrong size it should just error. This might be frustrating because you'd have to pad Buffers but I think these things should be strictly enforced. I really think we should include this in V5.

**Comment by alcuadrado** [comment]
> This might be frustrating because you'd have to pad Buffers

This is operation is not that common though. IMO, as long as the VM does it (does it?) we shouldn't need to worry about it being inconvenient.

**Comment by jochem-brouwer** [comment]
Ah okay, good point. We should throw if the key/value > 32 bytes and we should internally pad keys/values with zeros if they are < 32 bytes. I have no problem with that.

**Comment by holgerd77** [comment]
> Ah okay, good point. We should throw if the key/value > 32 bytes and we should internally pad keys/values with zeros if they are < 32 bytes. I have no problem with that.

Can you confirm this final summary from Jochem, not that there is some misinterpretation here which might lead to a differing implementation approach?

**Comment by holgerd77** [comment]
ah, this was: @alcuadrado 😄

**Comment by holgerd77** [comment]
@jochem-brouwer also I am getting a bit lost on this discussion. For `key` 32-byte length is actually already enforced. Could you state for the following `value` flavors what you think should happen? Also @alcuadrado?

- Empty Buffer (`Buffer.alloc(0)`)
- Non-32-byte zero Buffer (`Buffer.alloc(2)`)
- 32-byte zero Buffer (`Buffer.alloc(32)`)

Some external references on this discussion would also be helpful.

**Comment by jochem-brouwer** [comment]
Hey @holgerd77 , you are right - actually, 32-byte key length is enforced. 

I just did some tests and the results are very surprising. 

```javascript
const VM = require('../../dist/index').default // file is in packages/vm/tests/api/

let VM1 = new VM()
let VM2 = new VM()
let VM3 = new VM()

let VM4 = new VM()
let VM5 = new VM()

let addr = Buffer.from(("0").repeat(40), 'hex')
let key = Buffer.from(("0").repeat(64), 'hex')

let value1 =    "0000000000000000000000000000000000000000000000000000000000000100" // 32 bytes value
let value2 =    "00000000000000000000000000000000000000000000000000000000000001"   // should be same value if it is right-padded
let value3 =      "00000000000000000000000000000000000000000000000000000000000100" // should be same value if it is left-padded
let value_zero ="0000000000000000000000000000000000000000000000000000000000000000"

let addr_wrongLength = Buffer.from(("0").repeat(38), 'hex')

async function go() {
    await VM1.stateManager.putContractStorage(addr, key, value1)
    console.log((await VM1.stateManager.getStateRoot()).toString('hex'), "put 0x00..100 in the zero slot")
    await VM2.stateManager.putContractStorage(addr, key, value2)
    console.log((await VM2.stateManager.getStateRoot()).toString('hex'), "put 31-byte 0x00..01 in the zero slot")
    await VM3.stateManager.putContractStorage(addr, key, value3)
    console.log((await VM3.stateManager.getStateRoot()).toString('hex'), "put 31-byte 0x00..0100 in the zero slot")
    console.log((await VM4.stateManager.getStateRoot()).toString('hex'), "this is the default state root after instantiating a VM")
    await VM4.stateManager.putContractStorage(addr, key, value_zero)
    console.log((await VM4.stateManager.getStateRoot()).toString('hex'), "state root after setting slot 0 of address 0 to 32-byte 0")
    await VM5.stateManager.putContractStorage(addr_wrongLength, key, value_zero)
    console.log((await VM5.stateManager.getStateRoot()).toString('hex'), "state root after setting slot 0 of address 0 (19-bytes long instead of 20) to 32-byte 0")
}

go()
```

Output:

```
cca58677ce056fa4d93542acf4fbba53a80bbb2265e807b7d0c5494faf49692b put 0x00..100 in the zero slot
de79d721d8df7204785190abc005c419ab4c7f3354fbed10a71e27446b8f0094 put 31-byte 0x00..01 in the zero slot
4909338143fc0c5eacf1ea56c9e9e3d03bfe759c4bd1b43345f3ac2fe8200096 put 31-byte 0x00..0100 in the zero slot
56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421 this is the default state root after instantiating a VM
b8439cdd7eac1aa1ab3bd3a492dfaf71bf790f26d555574bc8716f640f4010e5 state root after setting slot 0 of address 0 to 32-byte 0
16d5b530c4e835fd0e7b80c0348e8b55f98a2b103a3a9d3513aff772a6815557 state root after setting slot 0 of address 0 (19-bytes long instead of 20) to 32-byte 0
```

Conclusions:

If an user puts storage and the Buffer length is wrong, then it is impossible to get the same state root, even if you consider right or left padding. It seems like values are not padded and these yield the wrong MPT state roots. (First 3 tests)

Putting a zero value into the zero slot changes the state root (I have to check this but I am fairly certain that this should not change the state root; the reason is that this zero-value is not saved and this is also why you get "gas refunds" if you clear a gas slot from a nonzero value to a zero value, to "clean up the trie").

If you put a value into a wrongly-padded address (19 byte zero address instead of the 20 byte zero address) you get a different state root.

This is a StateManager issue not a VM one, but I think the point is clear.

**Comment by jochem-brouwer** [comment]
Related (from yellow paper):

"The core of the trie, and its sole requirement in terms of the protocol specification, is to provide a single value that identifies a given set of key-value pairs, which may be either a 32-byte sequence or the empty byte sequence." (appendix D)

Internally we do not strictly ensure that the value is either empty (the "zero" value, the "empty" slot, etc.) or that it is 32 bytes. We should ensure that if we pass storage values around they are either empty (zero bytes) or 32 bytes.

**Comment by holgerd77** [comment]
Is this completely fixed by #870 now?

**Comment by jochem-brouwer** [comment]
It is fixed internally, but via the externally exposed API (`StateManager` `putContractStorage`) this is still possible. In all cases a zero buffer should not be stored in the MPT. `SecureTrie` is not explicitly the Ethereum MPT: it allows zero values to be stored. So we should change this in the `StateManager`: if you want to store a zero buffer, then don't store it.
